livekit-agents/livekit/agents/voice/room_io/_pre_connect_audio.py

```diff
@@ -0,0 +1,117 @@
+import asyncio
+import contextlib
+import time
+from collections.abc import Coroutine
+from dataclasses import dataclass, field
+from typing import Any, Callable
+
+from livekit import rtc
+
+from ..agent import logger, utils
+
+PRE_CONNECT_AUDIO_BUFFER_STREAM = "lk.agent.pre-connect-audio-buffer"
+PRE_CONNECT_AUDIO_ATTRIBUTE = "lk.agent.pre-connect-audio"
+
+
+@dataclass
+class _PreConnectAudioBuffer:
+    timestamp: float
+    frames: list[rtc.AudioFrame] = field(default_factory=list)
+
+
+_WaitPreConnectAudio = Callable[[rtc.RemoteParticipant], Coroutine[Any, Any, list[rtc.AudioFrame]]]
+
+
+class PreConnectAudioHandler:
+    def __init__(self, room: rtc.Room, *, timeout: float, max_delta_s: float = 1.0):
+        self._room = room
+        self._timeout = timeout
+        self._max_delta_s = max_delta_s
+
+        self._buffers: dict[str, asyncio.Future[_PreConnectAudioBuffer]] = {}
+        self._lock = asyncio.Lock()
+        self._tasks: set[asyncio.Task] = set()
+
+    def register(self):
+        def _handler(reader: rtc.ByteStreamReader, participant_id: str):
+            task = asyncio.create_task(self._read_audio_task(reader, participant_id))
+            self._tasks.add(task)
+            task.add_done_callback(self._tasks.discard)
+
+        if self._room.isconnected():
+            logger.warning("pre-connect audio handler registered after room is connected")
```

Comment:
let's move to the start method of room_io, and explain this is due to the pre_connect_audio=True